File: src/go-pdk/extism_pdk/extism_pdk.go
// Package extism_pdk provides a Go interface for developing Extism plugins
package extism_pdk

import (
	"encoding/json"
	"fmt"
)

// Memory operations - these are imported from the host environment
//
//export extism_input_length
func extism_input_length() uint64

//export extism_input_load
func extism_input_load(offset uint64, length uint64) uint64

//export extism_output_set
func extism_output_set(offset uint64, length uint64) uint64

//export extism_error_set
func extism_error_set(offset uint64, length uint64) uint64

//export extism_length
func extism_length(id uint64) uint64

//export extism_alloc
func extism_alloc(length uint64) uint64

//export extism_free
func extism_free(offset uint64)

//export extism_store_u8
func extism_store_u8(offset uint64, value uint8)

//export extism_store_u64
func extism_store_u64(offset uint64, value uint64)

//export extism_load_u8
func extism_load_u8(offset uint64) uint8

//export extism_load_u64
func extism_load_u64(offset uint64) uint64

// Host functions - these are functions provided by the host
//
//export extism_http_request
func extism_http_request(request uint64, request_length uint64) uint64

//export extism_http_status_code
func extism_http_status_code() uint64

//export extism_config_get
func extism_config_get(key uint64, key_length uint64) uint64

//export extism_var_get
func extism_var_get(key uint64, key_length uint64) uint64

//export extism_var_set
func extism_var_set(key uint64, key_length uint64, value uint64, value_length uint64) uint64

//export extism_log_info
func extism_log_info(msg uint64, msg_length uint64)

//export extism_log_debug
func extism_log_debug(msg uint64, msg_length uint64)

//export extism_log_warn
func extism_log_warn(msg uint64, msg_length uint64)

//export extism_log_error
func extism_log_error(msg uint64, msg_length uint64)

// Host is the main interface for interacting with the host environment
type Host struct{}

// GetInput returns the input data provided to the plugin
func (h Host) GetInput() []byte {
	length := extism_input_length()
	if length == 0 {
		return []byte{}
	}

	memory := make([]byte, length)
	ptr := extism_input_load(0, length)

	// Copy from host memory to Go slice
	for i := uint64(0); i < length; i++ {
		memory[i] = extism_load_u8(ptr + i)
	}

	return memory
}

// GetInputString returns the input data as a string
func (h Host) GetInputString() string {
	return string(h.GetInput())
}

// GetInputJSON unmarshals the input JSON into the provided interface
func (h Host) GetInputJSON(v interface{}) error {
	data := h.GetInput()
	return json.Unmarshal(data, v)
}

// SetOutput sets the output data for the plugin
func (h Host) SetOutput(data []byte) error {
	length := uint64(len(data))
	ptr := extism_alloc(length)

	// Copy from Go slice to host memory
	for i := uint64(0); i < length; i++ {
		extism_store_u8(ptr+i, data[i])
	}

	extism_output_set(ptr, length)
	extism_free(ptr)
	return nil
}

// SetOutputString sets the output string for the plugin
func (h Host) SetOutputString(s string) error {
	return h.SetOutput([]byte(s))
}

// SetOutputJSON marshals the provided interface to JSON and sets it as output
func (h Host) SetOutputJSON(v interface{}) error {
	data, err := json.Marshal(v)
	if err != nil {
		return err
	}
	return h.SetOutput(data)
}

// SetError sets an error message for the plugin
func (h Host) SetError(msg string) error {
	data := []byte(msg)
	length := uint64(len(data))
	ptr := extism_alloc(length)

	// Copy from Go slice to host memory
	for i := uint64(0); i < length; i++ {
		extism_store_u8(ptr+i, data[i])
	}

	extism_error_set(ptr, length)
	extism_free(ptr)
	return nil
}

// LogInfo logs an informational message
func (h Host) LogInfo(msg string) {
	data := []byte(msg)
	length := uint64(len(data))
	ptr := extism_alloc(length)

	// Copy from Go slice to host memory
	for i := uint64(0); i < length; i++ {
		extism_store_u8(ptr+i, data[i])
	}

	extism_log_info(ptr, length)
	extism_free(ptr)
}

// LogDebug logs a debug message
func (h Host) LogDebug(msg string) {
	data := []byte(msg)
	length := uint64(len(data))
	ptr := extism_alloc(length)

	// Copy from Go slice to host memory
	for i := uint64(0); i < length; i++ {
		extism_store_u8(ptr+i, data[i])
	}

	extism_log_debug(ptr, length)
	extism_free(ptr)
}

// LogWarn logs a warning message
func (h Host) LogWarn(msg string) {
	data := []byte(msg)
	length := uint64(len(data))
	ptr := extism_alloc(length)

	// Copy from Go slice to host memory
	for i := uint64(0); i < length; i++ {
		extism_store_u8(ptr+i, data[i])
	}

	extism_log_warn(ptr, length)
	extism_free(ptr)
}

// LogError logs an error message
func (h Host) LogError(msg string) {
	data := []byte(msg)
	length := uint64(len(data))
	ptr := extism_alloc(length)

	// Copy from Go slice to host memory
	for i := uint64(0); i < length; i++ {
		extism_store_u8(ptr+i, data[i])
	}

	extism_log_error(ptr, length)
	extism_free(ptr)
}

// HTTPRequest makes an HTTP request to the host
type HTTPRequest struct {
	Method  string            `json:"method"`
	URL     string            `json:"url"`
	Headers map[string]string `json:"headers,omitempty"`
	Body    string            `json:"body,omitempty"`
}

// HTTPResponse is the response from an HTTP request
type HTTPResponse struct {
	Status  int               `json:"status"`
	Headers map[string]string `json:"headers,omitempty"`
	Body    string            `json:"body"`
}

// HTTP makes an HTTP request
func (h Host) HTTP(req HTTPRequest) (*HTTPResponse, error) {
	data, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}

	length := uint64(len(data))
	ptr := extism_alloc(length)

	// Copy from Go slice to host memory
	for i := uint64(0); i < length; i++ {
		extism_store_u8(ptr+i, data[i])
	}

	resultPtr := extism_http_request(ptr, length)
	extism_free(ptr)

	if resultPtr == 0 {
		return nil, fmt.Errorf("HTTP request failed")
	}

	resultLength := extism_length(resultPtr)
	result := make([]byte, resultLength)

	// Copy from host memory to Go slice
	for i := uint64(0); i < resultLength; i++ {
		result[i] = extism_load_u8(resultPtr + i)
	}

	status := extism_http_status_code()

	var response HTTPResponse
	err = json.Unmarshal(result, &response)
	if err != nil {
		return nil, err
	}

	response.Status = int(status)
	return &response, nil
}

// GetConfig gets a configuration value by key
func (h Host) GetConfig(key string) string {
	data := []byte(key)
	length := uint64(len(data))
	ptr := extism_alloc(length)

	// Copy from Go slice to host memory
	for i := uint64(0); i < length; i++ {
		extism_store_u8(ptr+i, data[i])
	}

	resultPtr := extism_config_get(ptr, length)
	extism_free(ptr)

	if resultPtr == 0 {
		return ""
	}

	resultLength := extism_length(resultPtr)
	result := make([]byte, resultLength)

	// Copy from host memory to Go slice
	for i := uint64(0); i < resultLength; i++ {
		result[i] = extism_load_u8(resultPtr + i)
	}

	return string(result)
}

// GetVar gets a variable value by key
func (h Host) GetVar(key string) string {
	data := []byte(key)
	length := uint64(len(data))
	ptr := extism_alloc(length)

	// Copy from Go slice to host memory
	for i := uint64(0); i < length; i++ {
		extism_store_u8(ptr+i, data[i])
	}

	resultPtr := extism_var_get(ptr, length)
	extism_free(ptr)

	if resultPtr == 0 {
		return ""
	}

	resultLength := extism_length(resultPtr)
	result := make([]byte, resultLength)

	// Copy from host memory to Go slice
	for i := uint64(0); i < resultLength; i++ {
		result[i] = extism_load_u8(resultPtr + i)
	}

	return string(result)
}

// SetVar sets a variable value by key
func (h Host) SetVar(key string, value string) bool {
	keyData := []byte(key)
	keyLength := uint64(len(keyData))
	keyPtr := extism_alloc(keyLength)

	// Copy key from Go slice to host memory
	for i := uint64(0); i < keyLength; i++ {
		extism_store_u8(keyPtr+i, keyData[i])
	}

	valueData := []byte(value)
	valueLength := uint64(len(valueData))
	valuePtr := extism_alloc(valueLength)

	// Copy value from Go slice to host memory
	for i := uint64(0); i < valueLength; i++ {
		extism_store_u8(valuePtr+i, valueData[i])
	}

	result := extism_var_set(keyPtr, keyLength, valuePtr, valueLength)

	extism_free(keyPtr)
	extism_free(valuePtr)

	return result == 1
}

// CreateHost creates a new Host instance
func CreateHost() Host {
	return Host{}
}

File: src/python-pdk/extism_pdk.py
"""
Extism Plugin Development Kit (PDK) for Python

This module provides a Python interface for developing Extism plugins.
"""

import json
import sys
from typing import Any, Dict, List, Optional, Union

# Import WASM functions from the host
# Note: These are defined as external imports by the Extism runtime
def extism_input_length() -> int:
    """Get the length of the input data"""
    ...

def extism_input_load(offset: int, length: int) -> int:
    """Load input data into memory"""
    ...

def extism_output_set(offset: int, length: int) -> int:
    """Set the output data"""
    ...

def extism_error_set(offset: int, length: int) -> int:
    """Set an error message"""
    ...

def extism_length(id: int) -> int:
    """Get the length of a memory block"""
    ...

def extism_alloc(length: int) -> int:
    """Allocate memory"""
    ...

def extism_free(offset: int) -> None:
    """Free memory"""
    ...

def extism_store_u8(offset: int, value: int) -> None:
    """Store a byte in memory"""
    ...

def extism_load_u8(offset: int) -> int:
    """Load a byte from memory"""
    ...

def extism_http_request(request: int, request_length: int) -> int:
    """Make an HTTP request"""
    ...

def extism_http_status_code() -> int:
    """Get the HTTP status code from the last request"""
    ...

def extism_config_get(key: int, key_length: int) -> int:
    """Get a configuration value"""
    ...

def extism_var_get(key: int, key_length: int) -> int:
    """Get a variable value"""
    ...

def extism_var_set(key: int, key_length: int, value: int, value_length: int) -> int:
    """Set a variable value"""
    ...

def extism_log_info(msg: int, msg_length: int) -> None:
    """Log an info message"""
    ...

def extism_log_debug(msg: int, msg_length: int) -> None:
    """Log a debug message"""
    ...

def extism_log_warn(msg: int, msg_length: int) -> None:
    """Log a warning message"""
    ...

def extism_log_error(msg: int, msg_length: int) -> None:
    """Log an error message"""
    ...

class Memory:
    """
    Helper class for memory management
    """
    @staticmethod
    def alloc(data: bytes) -> int:
        """
        Allocate memory and store data
        
        Args:
            data: The data to store
            
        Returns:
            The memory offset
        """
        length = len(data)
        offset = extism_alloc(length)
        
        for i, b in enumerate(data):
            extism_store_u8(offset + i, b)
            
        return offset
    
    @staticmethod
    def free(offset: int) -> None:
        """
        Free memory
        
        Args:
            offset: The memory offset to free
        """
        extism_free(offset)
    
    @staticmethod
    def get(offset: int, length: int) -> bytes:
        """
        Get data from memory
        
        Args:
            offset: The memory offset
            length: The length of data to get
            
        Returns:
            The data as bytes
        """
        data = bytearray(length)
        for i in range(length):
            data[i] = extism_load_u8(offset + i)
        return bytes(data)


class Host:
    """
    The main interface for interacting with the host environment
    """
    
    def get_input(self) -> bytes:
        """
        Get the input data as bytes
        
        Returns:
            The input data
        """
        length = extism_input_length()
        if length == 0:
            return b""
        
        offset = extism_input_load(0, length)
        return Memory.get(offset, length)
    
    def get_input_string(self) -> str:
        """
        Get the input data as a string
        
        Returns:
            The input data as a string
        """
        try:
            return self.get_input().decode('utf-8')
        except UnicodeDecodeError:
            return ""
    
    def get_input_json(self) -> Any:
        """
        Get the input data as JSON
        
        Returns:
            The parsed JSON data
        """
        try:
            return json.loads(self.get_input_string())
        except json.JSONDecodeError:
            return None
    
    def set_output(self, data: bytes) -> None:
        """
        Set the output data
        
        Args:
            data: The output data
        """
        offset = Memory.alloc(data)
        extism_output_set(offset, len(data))
        Memory.free(offset)
    
    def set_output_string(self, data: str) -> None:
        """
        Set the output data as a string
        
        Args:
            data: The output string
        """
        self.set_output(data.encode('utf-8'))
    
    def set_output_json(self, data: Any) -> None:
        """
        Set the output data as JSON
        
        Args:
            data: The data to serialize as JSON
        """
        self.set_output_string(json.dumps(data))
    
    def set_error(self, error: str) -> None:
        """
        Set an error message
        
        Args:
            error: The error message
        """
        data = error.encode('utf-8')
        offset = Memory.alloc(data)
        extism_error_set(offset, len(data))
        Memory.free(offset)
    
    def log_info(self, message: str) -> None:
        """
        Log an informational message
        
        Args:
            message: The message to log
        """
        data = message.encode('utf-8')
        offset = Memory.alloc(data)
        extism_log_info(offset, len(data))
        Memory.free(offset)
    
    def log_debug(self, message: str) -> None:
        """
        Log a debug message
        
        Args:
            message: The message to log
        """
        data = message.encode('utf-8')
        offset = Memory.alloc(data)
        extism_log_debug(offset, len(data))
        Memory.free(offset)
    
    def log_warn(self, message: str) -> None:
        """
        Log a warning message
        
        Args:
            message: The message to log
        """
        data = message.encode('utf-8')
        offset = Memory.alloc(data)
        extism_log_warn(offset, len(data))
        Memory.free(offset)
    
    def log_error(self, message: str) -> None:
        """
        Log an error message
        
        Args:
            message: The message to log
        """
        data = message.encode('utf-8')
        offset = Memory.alloc(data)
        extism_log_error(offset, len(data))
        Memory.free(offset)
    
    def http_request(self, 
                     url: str, 
                     method: str = "GET", 
                     headers: Optional[Dict[str, str]] = None, 
                     body: Optional[str] = None) -> Dict[str, Any]:
        """
        Make an HTTP request
        
        Args:
            url: The URL to request
            method: The HTTP method
            headers: Optional HTTP headers
            body: Optional request body
            
        Returns:
            The HTTP response as a dictionary
        """
        request = {
            "url": url,
            "method": method
        }
        
        if headers:
            request["headers"] = headers
            
        if body:
            request["body"] = body
            
        request_json = json.dumps(request).encode('utf-8')
        request_offset = Memory.alloc(request_json)
        
        response_offset = extism_http_request(request_offset, len(request_json))
        Memory.free(request_offset)
        
        if response_offset == 0:
            return {"status": 0, "body": ""}
            
        response_length = extism_length(response_offset)
        response_data = Memory.get(response_offset, response_length)
        
        try:
            response = json.loads(response_data.decode('utf-8'))
            response["status"] = extism_http_status_code()
            return response
        except (UnicodeDecodeError, json.JSONDecodeError):
            return {"status": extism_http_status_code(), "body": ""}
    
    def get_config(self, key: str) -> str:
        """
        Get a configuration value
        
        Args:
            key: The configuration key
            
        Returns:
            The configuration value
        """
        key_bytes = key.encode('utf-8')
        key_offset = Memory.alloc(key_bytes)
        
        result_offset = extism_config_get(key_offset, len(key_bytes))
        Memory.free(key_offset)
        
        if result_offset == 0:
            return ""
            
        result_length = extism_length(result_offset)
        result_data = Memory.get(result_offset, result_length)
        
        try:
            return result_data.decode('utf-8')
        except UnicodeDecodeError:
            return ""
    
    def get_var(self, key: str) -> str:
        """
        Get a variable value
        
        Args:
            key: The variable key
            
        Returns:
            The variable value
        """
        key_bytes = key.encode('utf-8')
        key_offset = Memory.alloc(key_bytes)
        
        result_offset = extism_var_get(key_offset, len(key_bytes))
        Memory.free(key_offset)
        
        if result_offset == 0:
            return ""
            
        result_length = extism_length(result_offset)
        result_data = Memory.get(result_offset, result_length)
        
        try:
            return result_data.decode('utf-8')
        except UnicodeDecodeError:
            return ""
    
    def set_var(self, key: str, value: str) -> bool:
        """
        Set a variable value
        
        Args:
            key: The variable key
            value: The variable value
            
        Returns:
            Whether the operation was successful
        """
        key_bytes = key.encode('utf-8')
        key_offset = Memory.alloc(key_bytes)
        
        value_bytes = value.encode('utf-8')
        value_offset = Memory.alloc(value_bytes)
        
        result = extism_var_set(key_offset, len(key_bytes), value_offset, len(value_bytes))
        
        Memory.free(key_offset)
        Memory.free(value_offset)
        
        return result == 1


# Create a global host instance for easier access
host = Host() 